sdk/eventhub/event-hubs/src/eventHubReceiver.ts

```diff
@@ -676,153 +523,224 @@ export class EventHubReceiver extends LinkEntity {
-            `address "${address}" has been cancelled by the user.`;
-          // Cancellation is intentional so logging to 'info'.
-          logger.info(desc);
-        };
-
-        const rejectOnAbort = async (): Promise<void> => {
-          logOnAbort();
-          try {
-            await this.close();
-          } finally {
-            reject(new AbortError(StandardAbortMessage));
-          }
-        };
+    this._isReceivingMessages = true;
+    this._isStreaming = false;
 
-        // operation has been cancelled, so exit immediately
-        if (abortSignal && abortSignal.aborted) {
-          return rejectOnAbort();
-        }
+    const cleanupBeforeAbort = (): Promise<void> => {
+      logger.info(
+        `[${this._context.connectionId}] The request operation on the Receiver "${this.name}" with address "${this.address}" has been cancelled by the user.`
+      );
+      return this.close();
+    };
 
-        // updates the prefetch count so that the baseConsumer adds
-        // the correct number of credits to receive the same number of events.
-        const prefetchCount = Math.max(maxMessageCount - receivedEvents.length, 0);
-        if (prefetchCount === 0) {
-          return resolve(receivedEvents);
+    const tryOpenLink = (): Promise<void> =>
+      createAbortablePromise(
+        (resolve, reject) =>
+          this.initialize({
+            abortSignal,
+            timeoutInMs: getRetryAttemptTimeoutInMs(this.options.retryOptions),
+          })
+            .then(resolve)
+            .catch(reject),
+        {
+          abortSignal,
+          abortErrorMsg: StandardAbortMessage,
         }
+      );
 
-        logger.verbose(
-          "[%s] Receiver '%s', setting the prefetch count to %d.",
-          this._context.connectionId,
-          this.name,
-          prefetchCount
-        );
-
-        const cleanUpBeforeReturn = (): void => {
-          this.clearHandlers();
-          clearTimeout(timer);
-        };
+    /** The time to wait in ms before attempting to read from the queue */
+    const readIntervalWaitTimeInMs = 20;
 
-        const onAbort = (): void => {
-          clearTimeout(timer);
-          rejectOnAbort();
-        };
+    const retrieveEvents = (): Promise<ReceivedEventData[]> => {
+      const eventsToRetrieveCount = Math.max(maxMessageCount - this.queue.length, 0);
+      logger.verbose(
+        "[%s] Receiver '%s' already has %d events and wants to receive %d more events.",
+        this._context.connectionId,
+        this.name,
+        this.queue.length,
+        eventsToRetrieveCount
+      );
+      if (abortSignal?.aborted) {
+        cleanupBeforeAbort();
+        return Promise.reject(new AbortError(StandardAbortMessage));
+      }
+      return this._isClosed || this._context.wasConnectionCloseCalled
+        ? Promise.resolve(this.queue.splice(0))
+        : eventsToRetrieveCount === 0
+        ? Promise.resolve(this.queue.splice(0, maxMessageCount))
```

Comment:
Why do we stop at the max message count?

Reply:
The customer asked for up to maMessageCount so we honor that. Is there a reason not to?

Reply:
what's the point of receiving more then?

Reply:
In service-bus this would impact badly as the delivery count will be incremented, I know it's not the same for event-hubs. But would better ponder over this now than later.

Reply:
We potentially return more than `maxMessageCount` only in the case the receiver is closed and it already has `maxMessageCount` in the internal queue for some reason. I can't think of a case where that would happen for a typical `maxMessageCount`. Anyhow, I believe the intent here is to make sure the customer receives everything already received up to this point. Let me know if you feel strongly about it and I can reconsider this.